src/widgets/AutoCompletion/lib/useAutoCompletion.ts

```diff
@@ -0,0 +1,195 @@
+import { useCallback, useContext, useState } from 'react';
+
+import { appletModel } from '~/entities/applet';
+import { mapItemToRecord } from '~/entities/applet/model/mapper';
+import { AutoCompletionModel } from '~/features/AutoCompletion';
+import { SurveyContext, useAnswer, useSubmitAnswersMutations } from '~/features/PassSurvey';
+import { ActivityApiProxyService, ActivityDTO, AnswerPayload } from '~/shared/api';
+
+type Item = appletModel.ItemRecord;
+
+type CompleteActivityParams = {
+  activityId: string;
+  items: Item[];
+  userEvents: appletModel.UserEvent[];
+  isLastActivity: boolean;
+};
+
+export const useAutoCompletion = () => {
+  const context = useContext(SurveyContext);
+
+  const [activityName, setActivityName] = useState<string>(context.activity.name);
+
+  const completionState = AutoCompletionModel.useAutoCompletionRecord({
+    entityId: context.entityId,
+    eventId: context.eventId,
+  });
+
+  const { activitySuccessfullySubmitted } = AutoCompletionModel.useAutoCompletionStateManager();
+
+  const { getActivityProgress } = appletModel.hooks.useActivityProgress();
+
+  const { buildAnswer } = useAnswer();
+
+  const { submitAnswersAsync } = useSubmitAnswersMutations({
+    isPublic: !!context.publicAppletKey,
+  });
+
+  const isLastActivity = useCallback(
+    (activityId: string) => {
+      if (!completionState) {
+        throw new Error('[useAutoCompletion:isLastActivity] AutoCompletion state is not defined');
+      }
+
+      const lastIndex = completionState.activityIdsToSubmit.length - 1;
+
+      return completionState.activityIdsToSubmit[lastIndex] === activityId;
+    },
+    [completionState],
+  );
+
+  const fetchActivityById = useCallback(
+    async (id: string): Promise<ActivityDTO> => {
+      let activityDTO: ActivityDTO | undefined;
+
+      try {
+        const response = await ActivityApiProxyService.getActivityById(id, {
+          isPublic: !!context.publicAppletKey,
+        });
+
+        activityDTO = response.data.result;
+      } catch (error) {
+        console.error(error);
+        throw new Error(
+          `[CompletionContructService:getActivityById] Error while fetching activity by ID: ${id}`,
+        );
+      }
+
+      setActivityName(activityDTO.name);
+      return activityDTO;
```

Comment:
The function has two responsibilities instead of one:
1. It fetches data from API server and returns the result
2. It changes the React state

Please consider the following approach: 
Make this function pure, meaning that it should be stored separately, outside the hook `and` it should not have side effects (changing React state). The function depends on `publicAppletKey` so it could be easily refactored moved out.

Reply:
Done